text/0064-asset-garbage-collection.md

```diff
@@ -0,0 +1,175 @@
+
+* **Original Author(s):**: @eladb, @kaizen3031593
+* **Tracking Issue**: #64
+* **API Bar Raiser**: @nlynch
+
+The asset garbage collection CLI will identify and/or delete orphaned CDK assets,
+resulting in smaller bucket/repo size and less cost for customers.
+
+## Working Backwards
+
+**CHANGELOG**:
+
+- feat(cli): implement `cdk gc` for garbage collection of orphaned assets
+
+**Help**:
+
+```shell
+➜ cdk gc --help
+cdk gc [ENVIRONMENT...]
+
+Finds and deletes all orphaned S3 and ECR assets in the ENVIRONMENT
+
+Options:
+  -l, --list              lists, rather than deletes, all orphaned assets
+  -t, --type=[s3|ecr]     filters for type of asset
+
+Examples:
+  cdk gc
+  cdk gc aws://ACCOUNT/REGION
+  cdk gc --list --type=s3
+```
+
+**README:**
+
+The `cdk gc` command reduces the size of your s3 asset bucket and your ecr asset
+repository by identifying and deleting orphaned assets. The command will enter the
+specified environment or all environments in the CDK app and determine assets that are no
+longer referenced and can be safely deleted. You can also list orphaned assets using the
+`--list` option or filter for a specific type of asset using `--type`.
+
+**Usage:**
+
+This command will garbage collect all orphaned assets in all environments that belong to the current CDK app (if `cdk.json` exists):
+
+```shell
+cdk gc
+```
+
+This command will list all orphaned assets in all environments that belong to the current CDK app:
+
+```shell
+cdk gc --list
+```
+
+This command will garbage collect orphaned assets in the specified environment:
+
+```shell
+cdk gc aws://ACCOUNT/REGION
+```
+
+This command will list orphaned s3 assets in the specified environment. The options for `--type` are `s3` or `ecr`:
+
+```shell
+cdk gc aws://ACOUNT/REGION --type=s3
+```
+
+---
+
+#
+
+Ticking the box below indicates that the public API of this RFC has been
+signed-off by the API bar raiser (the `api-approved` label was applied to the
+RFC pull request):
+
+```
+[ ] Signed-off by API Bar Raiser @xxxxx
+```
+
+## Public FAQ
+
+> This section should include answers to questions readers will likely ask about
+> this release. Similar to the "working backwards", this section should be
+> written in a language as if the feature is now released.
+>
+> The template includes a some common questions, feel free to add any questions
+> that might be relevant to this feature or omit questions that you feel are not
+> applicable.
+
+### What are we launching today?
+
+The `cdk gc` command features, with support for garbage collection of orphaned s3 and ecr
+assets.
+
+### Why should I use this feature?
+
+Currently orphaned assets are left in the s3 bucket or ecr repository and contribute
+additional cost for customers. This feature provides a swift way to identify and delete
+unutilized assets.
+
+### How does the command identify orphaned assets?
+
+I have not settled on a plan, so feel free to suggest. My initial thoughts are that at
```

Comment:
Since this does not happen during deployment, I think `gc` should be able to look at all the deployed stacks in the environment (with a healthy state) and trace back which assets are being referenced by these stacks. The remaining question is how do you know which assets are referenced by a CDK stack - is it based on some heuristics (e.g. search for a string pattern that represents an asset location) or based on explicit reporting (e.g. the stack provides metadata that indicates which assets are being used by it).

Reply:
I put this into the "high-level project plan" section and more or less copied these two suggestions as potential solutions so we can discuss further there.

Reply:
"project plan" refers to plan for executing the implementation. This probably belongs to the high level design section